### PR TITLE
Upgrade to GHC 8.10 and use Ubuntu 20.04 as base image for haskell docker images

### PR DESCRIPTION
**Pull request by fisx**
As the only bindists for GHC 8.10.7 on Alpine Linux are for integer-simple and migrating to integer-simple is some work, this PR swaps out the Alpine docker images with Ubuntu 20.04 based docker images. As Ubuntu is a much more widely used distribution among haskellers, we hope for better support from community for this.

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] **changelog.d** contains the following bits of information:
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

**Comment by mdimjasevic** [comment]
@fisx , is this still a draft?

**Comment by fisx** [comment]
yes.  i was hoping this would be easy, but it wasn't.  don't remember what i got stuck with.

**Comment by fisx** [comment]
The above error was with amazonka from hackage.  Now I am running `make` on 62715875cddda4b931022c7baccead1a90346625, and stack 2.3.1 is hanging here:

```
Cloning 412172d8c28906591f01576a78792de7c34cc3eb from https://github.com/wireapp/amazonka
```

stack-2.7.3 works better.

**Comment by fisx** [comment]
```
cryptohash                       > /tmp/stack-0a980421e8d42b9e/cryptohash-0.11.9/cbits/sha512.c:235:29: error:          
cryptohash                       >      warning: ‘%d’ directive writing between 1 and 11 bytes into a region of size 4 [-Wformat-overflow=]
cryptohash                       >       235 |   i = sprintf(buf, "SHA-512/%d", t);                                     
cryptohash                       >           |                             ^~                                           
cryptohash                       >     |                                                                                
cryptohash                       > 235 |                 i = sprintf(buf, "SHA-512/%d", t);                             
cryptohash                       >     |                             ^                                                  
cryptohash                       >                                                                                      
cryptohash                       > /tmp/stack-0a980421e8d42b9e/cryptohash-0.11.9/cbits/sha512.c:235:20: error:          
cryptohash                       >      note: directive argument in the range [-2147483648, 511]                        
cryptohash                       >       235 |   i = sprintf(buf, "SHA-512/%d", t);                                     
cryptohash                       >           |                    ^~~~~~~~~~~~                                          
cryptohash                       >     |                                                                                
cryptohash                       > 235 |                 i = sprintf(buf, "SHA-512/%d", t);                             
cryptohash                       >     |                    ^                                                           
cryptohash                       > In file included from /nix/store/f8ahd6cxx9ia4fzdiyczmhcv3c2qmd42-glibc-2.32-37-dev/include/stdio.h:866,
cryptohash                       >                                                                                      
cryptohash                       > /tmp/stack-0a980421e8d42b9e/cryptohash-0.11.9/                 from cbits/sha512.c:197:0: error: 
cryptohash                       >                                                                                      
cryptohash                       > /tmp/stack-0a980421e8d42b9e/cryptohash-0.11.9//nix/store/f8ahd6cxx9ia4fzdiyczmhcv3c2qmd42-glibc-2.32-37-dev/include/bits/stdio2.h:38:10: error:
cryptohash                       >      note: ‘__builtin___sprintf_chk’ output between 10 and 20 bytes into a destination of size 12
cryptohash                       >        38 |   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,          
cryptohash                       >           |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~          
cryptohash                       >        39 |       __bos (__s), __fmt, __va_arg_pack ());                             
cryptohash                       >           |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                              
cryptohash                       >    |                                                                                 
cryptohash                       > 38 |   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,                 
cryptohash                       >    |          ^                                                                      
```

**Comment by fisx** [comment]
(maybe try cryptohash head from github main branch?)

**Comment by fisx** [comment]
```
amazonka-ses                     > [17 of 75] Compiling Network.AWS.SES.SendBounce          
                                                                                            
Warning: Failed to decode module interface:
         /home/mf/src/wire-server/libs/types-common-aws/.stack-work/dist/x86_64-linux-nix-tinfo6/Cabal-3.2.1.0/build/Util/Test/SQS.hi
         Decoding failure: Invalid magic: e49ceb0f
```

**Comment by fisx** [comment]
warning: we may have to use stack-2.7.3 or whatever's up to date now.

**Comment by akshaymankar** [comment]
Please ignore the main PR pipeline, it will not work with such big changes. There is a separate pipeline for this branch here: https://bridge-ie-concourse.zinfra.io/teams/main/pipelines/wire-server-ghc8107

File: build/ubuntu/Dockerfile.builder [hunk 1]
```diff
@@ -18,9 +17,9 @@ RUN set -x && \
     cd /wire-server && \
     stack update && \
     echo "allow-different-user: true" >> /root/.stack/config.yaml && \
-    stack build --haddock --dependencies-only haskell-src-exts && \
-    stack build --haddock --no-haddock-hyperlink-source haskell-src-exts && \
-    stack build --pedantic --haddock --test --no-run-tests --bench --no-run-benchmarks --dependencies-only -j${THREADS} && \
+    stack build --dependencies-only haskell-src-exts && \
+    stack build haskell-src-exts && \
+    stack build --pedantic --test --no-run-tests --bench --no-run-benchmarks --dependencies-only -j${THREADS} && \
```
Comment (fisx) [inline build/ubuntu/Dockerfile.builder:22]: why does haddock go away here (sorry, didn't go through the >100 commits individually for context).

File: libs/types-common/src/Data/Text/Ascii.hs [hunk 2]
```diff
@@ -339,9 +339,7 @@ encodeBase16 = unsafeFromByteString . B16.encode
 -- | Decode a text containing only hex characters.
 -- Decoding only succeeds if the text is a multiple of 2 bytes in length.
 decodeBase16 :: AsciiBase16 -> Maybe ByteString
-decodeBase16 t = case B16.decode (toByteString' t) of
-  (b, r) | r == mempty -> Just b
-  (_, _) -> Nothing
+decodeBase16 t = either (const Nothing) Just (B16.decode (toByteString' t))
```
Comment (fisx) [inline libs/types-common/src/Data/Text/Ascii.hs:342]: Now I want to remove this function and use plain `B16.decode` instead...  (but not here!)

File: libs/wire-api/src/Wire/API/Arbitrary.hs [hunk 3]
```diff
@@ -79,18 +80,22 @@ instance
...
 
 customSizedOpts :: CustomSizedOpts
 customSizedOpts =
   Generic.setGenerators
-    (Generic.Gen1 listOf' :+ Generic.Gen1 list1Of' :+ ())
+    (Generic.Gen1 listOf' :+ Generic.Gen1 nonEmptyListOf' :+ ())
     Generic.sizedOpts
 
 list1Of' :: Gen a -> Gen (List1 a)
 list1Of' g = list1 <$> g <*> Generic.listOf' g
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Arbitrary.hs:94]: `list1Of'` is now dead code.

File: nix/sources.json [hunk 4]
```diff
@@ -17,10 +17,10 @@
         "homepage": "https://github.com/NixOS/nixpkgs",
         "owner": "NixOS",
         "repo": "nixpkgs",
-        "rev": "8e1891d5b8d0b898db8890ddab73141f0cd3c2bc",
-        "sha256": "0a767mn0nfp4qnklsvs6bnc0vng4nc3ch566nmrz18ypk67z4zz0",
+        "rev": "d9896c1f8e419417ea8fc4b7e0e98ca4fc50ba8d",
+        "sha256": "1ysdab38jpjjvw5vc6qg1bns2f0hkwasww871hfhy9d9kxd6kaly",
```
Comment (fisx) [inline nix/sources.json:21]: should the niv, nixpkgs bump also be mentioned in the release notes?

File: services/gundeck/src/Gundeck/Monad.hs [hunk 5]
```diff
@@ -70,15 +70,10 @@ newtype Gundeck a = Gundeck
       MonadCatch,
       MonadMask,
       MonadReader Env,
-      MonadClient
+      MonadClient,
+      MonadUnliftIO
     )
 
-instance MonadUnliftIO Gundeck where
```
Comment (fisx) [inline services/gundeck/src/Gundeck/Monad.hs:76]: where are these coming from now?

File: tools/db/move-team/src/Schema.hs [hunk 6]
```diff
@@ -1224,7 +1224,7 @@ importSparUser Env {..} path = do
   if exists
     then do
       putStrLn $ "Importing " <> path <> " to " <> "spar.user"
-      withBinaryFile path ReadMode $ \handle -> do
```
Comment (fisx) [inline tools/db/move-team/src/Schema.hs:1227]: I prefer to find hlint stuff in separate PRs (but maybe that was hard because your tooling wouldn't let go of these easily?)

File: nix/sources.nix [hunk 7]
```diff
@@ -6,52 +6,63 @@ let
...
-  fetch_tarball = pkgs: spec:
-    if spec.builtin or true then
-      builtins_fetchTarball { inherit (spec) url sha256; }
-    else
-      pkgs.fetchzip { inherit (spec) url sha256; };
+  fetch_file = pkgs: name: spec:
+    let
+      name' = sanitizeName name + "-src";
+    in
+      if spec.builtin or true then
```
Comment (fisx) [inline nix/sources.nix:13]: you sure about this?

File: build/ubuntu/Dockerfile.intermediate [hunk 8]
```diff
@@ -0,0 +1,18 @@
+ARG builder=quay.io/wire/ubuntu20-builder
```
Comment (fisx) [inline build/ubuntu/Dockerfile.intermediate:1]: why did you remove the docs comments here?  (just curious)

File: build/ubuntu/Dockerfile.prebuilder [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+# get static version of Haskell Stack and use system ghc by default
+ARG STACK_VERSION=2.7.3
+RUN curl -sSfL https://github.com/commercialhaskell/stack/releases/download/v${STACK_VERSION}/stack-${STACK_VERSION}-linux-x86_64-static.tar.gz \
+    | tar --wildcards -C /usr/local/bin --strip-components=1 -xzvf - '*/stack' && chmod 755 /usr/local/bin/stack && \
+    stack config set system-ghc --global true
+
+ARG GHC_VERSION=8.10.7
+RUN curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org \
+  | BOOTSTRAP_HASKELL_NONINTERACTIVE=1 BOOTSTRAP_HASKELL_GHC_VERSION=${GHC_VERSION} sh
```
Comment (fisx) [inline build/ubuntu/Dockerfile.prebuilder:70]: why don't you install stack via ghcup?  seems easier and more robust?  (probably because it wasn't done before?)

**Review by fisx — COMMENTED** [review]
@akshaymankar feel free to approve it yourself; I can't since I started the PR :)

(small commit resolving the conflicts coming up.)

File: nix/sources.nix (hunk 7 above)
Comment (fisx) [reply nix/sources.nix:13]: never mind, this is generated, not our problem.  :)

**Comment by fisx** [comment]
last commit is wrong, in order for direnv to pick up the changes you need to leave the directory and come back.  i'll try again when i'm dong building.

File: Makefile [hunk 10]
```diff
@@ -189,39 +189,39 @@ i-%:
 .PHONY: docker-prebuilder
 docker-prebuilder:
 	# `docker-prebuilder` needs to be built or pulled only once (unless native dependencies change)
-	$(MAKE) -C build/alpine prebuilder
+	$(MAKE) -C build/ubuntu prebuilder
```
[the remark this answers is not in the document]
Comment (akshaymankar) [reply Makefile:192]: This is also intentional, that container just `echo`s sometime to a file.

File: build/ubuntu/Dockerfile.builder (hunk 1 above)
Comment (akshaymankar) [reply build/ubuntu/Dockerfile.builder:22]: It is not needed and was making the build slow. I guess this would mean hoogle builds are slightly slow, but I think that is fine.

File: nix/sources.json (hunk 4 above)
Comment (akshaymankar) [reply nix/sources.json:21]: It is only used in development dependencies. I will a note to the internal changes.

File: tools/db/move-team/src/Schema.hs (hunk 6 above)
Comment (akshaymankar) [reply tools/db/move-team/src/Schema.hs:1227]: Yeah, picking only a few suggestions was hard, and applying all was a good thing, so I applied all in this file.

File: services/gundeck/src/Gundeck/Monad.hs (hunk 5 above)
Comment (akshaymankar) [reply services/gundeck/src/Gundeck/Monad.hs:76]: It is being derived two lines above.

File: build/ubuntu/Dockerfile.prebuilder (hunk 9 above)
Comment (akshaymankar) [reply build/ubuntu/Dockerfile.prebuilder:70]: Oh right, ghcup can install stack too!

**Comment by akshaymankar** [comment]
> last commit is wrong, in order for direnv to pick up the changes you need to leave the directory and come back. i'll try again when i'm dong building.

Hmm, we need to add a bunch of `watch_file` commands to the `.envrc`.

File: build/ubuntu/Dockerfile.intermediate (hunk 8 above)
Comment (akshaymankar) [reply build/ubuntu/Dockerfile.intermediate:8]: Added them back.

File: build/ubuntu/Dockerfile.prebuilder (hunk 9 above)
Comment (akshaymankar) [reply build/ubuntu/Dockerfile.prebuilder]: Alright, done this. Thanks for the idea!

**Comment by fisx** [comment]
any other blockers?  i say :shipit:, but you'll have to formally approve since i created the PR.

**Comment by akshaymankar** [comment]
> any other blockers? i say :shipit:, but you'll have to formally approve since i created the PR.

I am planning to :shipit: on Monday, because I suspect I will have to baby sit some pipeline changes to be made in order and I don't want to leave things broken when I am not around.

**Review by akshaymankar — APPROVED** [review]
On behalf of @fisx
